File: dplyr_tpc/h04.R
tpc_h04 <- function(input_func, collect_func = dplyr::collect) {
  l <- input_func("lineitem") %>%
    select(l_orderkey, l_commitdate, l_receiptdate) %>%
    filter(l_commitdate < l_receiptdate) %>%
    select(l_orderkey)

  o <- input_func("orders") %>%
    select(o_orderkey, o_orderdate, o_orderpriority) %>%
    # kludge: filter(o_orderdate >= "1993-07-01", o_orderdate < "1993-07-01" + interval '3' month) %>%
    filter(o_orderdate >= as.Date("1993-07-01"), o_orderdate < as.Date("1993-10-01")) %>%
    select(o_orderkey, o_orderpriority)

  # distinct after join, tested and indeed faster
  lo <- inner_join(l, o, by = c("l_orderkey" = "o_orderkey")) %>%
    distinct() %>%
    select(o_orderpriority)

  lo %>%
    group_by(o_orderpriority) %>%
    summarise(order_count = n()) %>%
    ungroup() %>%
    arrange(o_orderpriority) %>%
    collect_func()
}

File: ibis_tpc/h04.py
"Order Priority Checking Query (Q4)"

import ibis


def tpc_h04(con, DATE="1993-07-01"):
    orders = con.table("orders")
    lineitem = con.table("lineitem")
    cond = (lineitem.l_orderkey == orders.o_orderkey) & (
        lineitem.l_commitdate < lineitem.l_receiptdate
    )
    q = orders.filter(
        [
            cond.any(),
            orders.o_orderdate.cast("date") >= ibis.date(DATE),
            orders.o_orderdate.cast("date") < ibis.date(DATE)
            + ibis.interval(months=3),
        ]
    )
    q = q.group_by([orders.o_orderpriority])
    q = q.aggregate(order_count=orders.count())
    q = q.order_by([orders.o_orderpriority])
    return q
